### ISSUE TITLE
Add 'addCheck' method to CommandManager

### ISSUE THREAD

**Issue by bramhaag** [aikar/commands#58]
In my opinion, the main problem with command frameworks is that you never have full control over them, which can result in either having to do some hacky methods to get it to do what you want, or just not using the framework but your own solution.

Issue #57 is a good example of this. If you want to add a default cooldown/per command cooldown, you'll have to modify ACF for it to work. My idea is to add a method to the CommandManager, something along the lines of `CommandManager#addCheck(Check)`, where `Check` is a functional interface which looks like this:
```java
interface Check {
    boolean check(Sender sender, String name, String[] args, Method method);
}
```

Again, using Issue #57 as an example, you can do this:
```java
myCommandManager.addCheck((sender, name, args, method) -> {
    boolean hasCooldown; //check if sender has cooldown
    if(hasCooldown) return false; //Command will not be executed
    
    int cooldown = method.getAnnotation(CommandCooldown.class).value(); //get cooldown (or default cooldown)
    //set cooldown for sender
   return true; //Command will be executed
});
```

**Comment by TryHardDood** [comment]
This would be cool.

**Comment by Williambraecky** [comment]
It'd probably be more logical to add it per command instead of per commandmanager, or even use an annotation with strings to identify the cheks to make it per subcommand compatible as you don't need the checks for every command in the command manager for example.
But yes somethings like this is a good idea and more flexible then just #57

**Comment by MiniDigger** [comment]
well, if you do it like in the example with an annotation, it would be logical to do it on a on command manager basis

**Comment by Williambraecky** [comment]
That part was thought before I thought of the annotation part lol.
It could be something like the resolver with .addCheck("checkName", whatever -> {});

**Comment by MiniDigger** [comment]
and have a special annotation @CommandChecks({"check1","check2"} so that you don't have to create bazillion of annotations.

meh, that would mean loosing the value of the annotation as config tho.
maybe @CommandCheck(name="checkname", value="someval")?

**Comment by chickeneer** [comment]
https://github.com/aikar/commands/blob/f900a739df7bfc0c1b6779be196541937209ddaf/core/src/main/java/co/aikar/commands/BaseCommand.java#L415

In the current version of ACF, one can override this method from BaseCommand to accomplish some of these ideas.
The method is called here: https://github.com/aikar/commands/blob/8168122241d56fc6278ef4ab2197b0ef5ddd1dc0/core/src/main/java/co/aikar/commands/RegisteredCommand.java#L133

(Just giving the context for how this would fit into existing code).

**Comment by Stawlker** [comment]
Yes, I agree with this.

**Comment by bramhaag** [comment]
Maybe having two types of checks would be better, global checks and per-command checks. So, `CommandManager#addCheck(String name, Check check)` and `CommandManager#addGlobalCheck(Check check)`. 

Per-command checks can be added to commands using the `@CommandChecks` annotations like @MiniDigger suggested, and maybe an annotation which disables the global checks, `@DisableGlobalChecks`, would also be useful.

**Comment by aikar** [comment]
I'm trying to move away from Override style stuff. Precommand is close, but I am working on something for this.

**Comment by Proximyst** [comment]
Has this died? Would love to see this, even though it would not be harder to manually do than something like follows:

```java
@Override
public boolean canExecute(CommandIssuer sender, RegisteredCommand<?> command) {
  if (!super.canExecute(sender, command)) return false;
  if (!(sender instanceof Player)) return true;
  boolean hasCooldown = getCooldown((Player) sender);
  if (!hasCooldown) {
    applyCooldown((Player) sender);
    return true;
  }
  // check cooldown time to override hasCooldown
  if (!hasCooldown) {
    applyCooldown((Player) sender;
    return true;
  }
  return false;
}
```

**Comment by aikar** [comment]
Conditions landed in 40eaa9ea07a4bf9ded34b251839d71c49182fb8f
Documentation: https://github.com/aikar/commands/wiki/Command-Conditions

Managing the state behind cooldowns can be done in your plugin.
You can just register 
```java
manager.getConditions().addCondition("cooldown", c -> { 
    Integer cooldown = c.getConfig("time", 5); // default 5s cooldown
    // .. validate cooldown
});

Then @Conditions("cooldown:time=15") // 15 second cooldown
```
